### PR TITLE
TLS Cipher 2a: tickets: use PSA for protection

### LINKED ISSUE #5203: TLS Cipher 2a: tickets: use PSA for protection
In `ssl_ticket.c`, use `psa_aead` instead of `mbedtls_cipher` in order to protect tickets when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

- [x] In `struct mbedtls_ssl_ticket_key` replace* the cipher context with two fields with types `mbedtls_svc_key_id_t` and `psa_algorithm_t`, and adapt `mbedtls_ssl_ticket_free()` accordingly. 
- [x] In `ssl_ticket_gen_key()`, set appropriate attributes (will need to use the algorithm now stored in the context) and import the the random bytes with `psa_import_key()`. (Compared to using `psa_generate_key()`, this allows for better code sharing between PSA and non-PSA, and is also preferable if we even want to enable PSA builds without `ENTROPY_C`, see https://github.com/ARMmbed/mbedtls/issues/5156#issuecomment-984662340)
- [x] In `mbedtls_ssl_ticket_setup()`, translate the `cipher` to the appropriate algorithm using the utility function created in #5182  and store that in the context before calling `ssl_ticket_gen_key()` - the calls to `mbedtls_cipher_setup()` don't make sense with PSA and can be simply removed.
- [x] In `mbedtls_ssl_ticket_write()`, instead of `mbedtls_cipher_auth_encrypt_ext()` call `psa_aead_encrypt()`.
- [x] In `mbedtls_ssl_ticket_parse()`, instead of `mbedtls_cipher_auth_decrypt_ext()` call `psa_aead_decrypt()`.
- [x] NEW: also handle the `mbedtls_ssl_ticket_rotate()` than was added by #5501
 
*Note: this describes the final result; in terms of execution, it may be desirable to first add the new fields (keeping the existing ones), then modify the functions gradually, and finally remove the old fields when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

Depends on: #5181 for tests, #5182 for translation function.

### PR DESCRIPTION
**Pull request by gabor-mezei-arm**
## Description
In `ssl_ticket.c`, use `psa_aead` instead of `mbedtls_cipher` in order to protect tickets when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

Resolve #5203

### CODE REVIEW

File: library/ssl_ticket.c [hunk 1]
```diff
@@ -83,11 +87,28 @@ static int ssl_ticket_gen_key( mbedtls_ssl_ticket_context *ctx,
     if( ( ret = ctx->f_rng( ctx->p_rng, buf, sizeof( buf ) ) ) != 0 )
         return( ret );
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+
```
Comment (superna9999) [inline library/ssl_ticket.c:91]:
```suggestion
```

File: library/ssl_ticket.c [hunk 2]
```diff
@@ -83,11 +87,28 @@ static int ssl_ticket_gen_key( mbedtls_ssl_ticket_context *ctx,
...
+                             PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT );
+    psa_set_key_algorithm( &attributes, key->alg );
+    psa_set_key_type( &attributes, key->key_type );
+    psa_set_key_bits( &attributes, key->key_bits );
+
+    ret = psa_ssl_status_to_mbedtls(
+            psa_import_key( &attributes, buf,
+                            PSA_BITS_TO_BYTES( key->key_bits ),
+                            &key->key ) );
+
```
Comment (superna9999) [inline library/ssl_ticket.c:102]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:102]:
```suggestion
```

File: library/ssl_ticket.c [hunk 3]
```diff
@@ -83,11 +87,28 @@ static int ssl_ticket_gen_key( mbedtls_ssl_ticket_context *ctx,
...
+#else
+
```
Comment (superna9999) [inline library/ssl_ticket.c:104]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:104]:
```suggestion
```

File: library/ssl_ticket.c [hunk 4]
```diff
@@ -83,11 +87,28 @@ static int ssl_ticket_gen_key( mbedtls_ssl_ticket_context *ctx,
...
     /* With GCM and CCM, same context can encrypt & decrypt */
     ret = mbedtls_cipher_setkey( &key->ctx, buf,
                                  mbedtls_cipher_get_key_bitlen( &key->ctx ),
                                  MBEDTLS_ENCRYPT );
```
Comment (superna9999) [inline library/ssl_ticket.c:109]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:109]:
```suggestion
```

File: library/ssl_ticket.c [hunk 5]
```diff
@@ -114,6 +139,14 @@ static int ssl_ticket_update_keys( mbedtls_ssl_ticket_context *ctx )
 
         ctx->active = 1 - ctx->active;
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+
```
Comment (superna9999) [inline library/ssl_ticket.c:143]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:143]:
```suggestion
```

File: library/ssl_ticket.c [hunk 6]
```diff
@@ -114,6 +139,14 @@ static int ssl_ticket_update_keys( mbedtls_ssl_ticket_context *ctx )
...
+        if( ( ret = psa_ssl_status_to_mbedtls(
+                        psa_destroy_key( ctx->keys[ctx->active].key ) ) ) != 0 )
+            return( ret );
+
```
Comment (superna9999) [inline library/ssl_ticket.c:147]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:147]:
```suggestion
```

File: library/ssl_ticket.c [hunk 7]
```diff
@@ -131,16 +164,44 @@ int mbedtls_ssl_ticket_rotate( mbedtls_ssl_ticket_context *ctx,
...
     const int bitlen = mbedtls_cipher_get_key_bitlen( &key->ctx );
-    int ret;
+#endif
+
     if( nlength < TICKET_KEY_NAME_BYTES || klength * 8 < (size_t)bitlen )
         return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
 
-    /* With GCM and CCM, same context can encrypt & decrypt */
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+
```
Comment (superna9999) [inline library/ssl_ticket.c:180]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:180]:
```suggestion
```

File: library/ssl_ticket.c [hunk 8]
```diff
@@ -131,16 +164,44 @@ int mbedtls_ssl_ticket_rotate( mbedtls_ssl_ticket_context *ctx,
...
+    psa_set_key_algorithm( &attributes, key->alg );
+    psa_set_key_type( &attributes, key->key_type );
+    psa_set_key_bits( &attributes, key->key_bits );
+
+    if( ( ret = psa_ssl_status_to_mbedtls(
+                    psa_import_key( &attributes, k,
+                                    PSA_BITS_TO_BYTES( key->key_bits ),
+                                    &key->key ) ) ) != 0 )
+        return( ret );
+
```
Comment (superna9999) [inline library/ssl_ticket.c:196]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:196]:
```suggestion
```

File: library/ssl_ticket.c [hunk 9]
```diff
@@ -131,16 +164,44 @@ int mbedtls_ssl_ticket_rotate( mbedtls_ssl_ticket_context *ctx,
...
+#else
+
```
Comment (superna9999) [inline library/ssl_ticket.c:198]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:198]:
```suggestion
```

File: library/ssl_ticket.c [hunk 10]
```diff
@@ -131,16 +164,44 @@ int mbedtls_ssl_ticket_rotate( mbedtls_ssl_ticket_context *ctx,
...
     ret = mbedtls_cipher_setkey( &key->ctx, k, bitlen, MBEDTLS_ENCRYPT );
     if( ret != 0 )
         return( ret );
```
Comment (superna9999) [inline library/ssl_ticket.c:202]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:202]:
```suggestion
```

File: library/ssl_ticket.c [hunk 11]
```diff
@@ -178,27 +246,29 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
...
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
-    ret = mbedtls_cipher_setup_psa( &ctx->keys[0].ctx,
-                                    cipher_info, TICKET_AUTH_TAG_BYTES );
-    if( ret != 0 && ret != MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE )
-        return( ret );
-    /* We don't yet expect to support all ciphers through PSA,
-     * so allow fallback to ordinary mbedtls_cipher_setup(). */
-    if( ret == MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE )
-#endif /* MBEDTLS_USE_PSA_CRYPTO */
+
```
Comment (superna9999) [inline library/ssl_ticket.c:249]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:249]:
```suggestion
```

File: library/ssl_ticket.c [hunk 12]
```diff
@@ -178,27 +246,29 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
...
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    ctx->keys[0].alg = alg;
+    ctx->keys[0].key_type = key_type;
+    ctx->keys[0].key_bits = key_bits;
+
+    ctx->keys[1].alg = alg;
+    ctx->keys[1].key_type = key_type;
+    ctx->keys[1].key_bits = key_bits;
+
```
Comment (superna9999) [inline library/ssl_ticket.c:261]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:261]:
```suggestion
```

File: library/ssl_ticket.c [hunk 13]
```diff
@@ -178,27 +246,29 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
...
+#else
+
```
Comment (superna9999) [inline library/ssl_ticket.c:263]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:263]:
```suggestion
```

File: library/ssl_ticket.c [hunk 14]
```diff
@@ -178,27 +246,29 @@ int mbedtls_ssl_ticket_setup( mbedtls_ssl_ticket_context *ctx,
...
 
-#if defined(MBEDTLS_USE_PSA_CRYPTO)
-    ret = mbedtls_cipher_setup_psa( &ctx->keys[1].ctx,
-                                    cipher_info, TICKET_AUTH_TAG_BYTES );
-    if( ret != 0 && ret != MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE )
-        return( ret );
-    if( ret == MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE )
-#endif /* MBEDTLS_USE_PSA_CRYPTO */
     if( ( ret = mbedtls_cipher_setup( &ctx->keys[1].ctx, cipher_info ) ) != 0 )
         return( ret );
```
Comment (superna9999) [inline library/ssl_ticket.c:269]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:269]:
```suggestion
```

File: library/ssl_ticket.c [hunk 15]
```diff
@@ -275,16 +345,25 @@ int mbedtls_ssl_ticket_write( void *p_ticket,
...
                     /* Additional data: key name, IV and length */
                     key_name, TICKET_ADD_DATA_LEN,
                     state, clear_len,
                     state, end - state, &ciph_len,
                     TICKET_AUTH_TAG_BYTES ) ) != 0 )
+#endif /* MBEDTLS_USE_PSA_CRYPTO */
     {
         goto cleanup;
     }
+
```
Comment (superna9999) [inline library/ssl_ticket.c:366]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:366]:
```suggestion
```

File: library/ssl_ticket.c [hunk 16]
```diff
@@ -418,9 +505,18 @@ int mbedtls_ssl_ticket_parse( void *p_ticket,
  */
 void mbedtls_ssl_ticket_free( mbedtls_ssl_ticket_context *ctx )
 {
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+
```
Comment (superna9999) [inline library/ssl_ticket.c:509]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:509]:
```suggestion
```

File: library/ssl_ticket.c [hunk 17]
```diff
@@ -418,9 +505,18 @@ int mbedtls_ssl_ticket_parse( void *p_ticket,
...
+    psa_destroy_key( ctx->keys[0].key );
+    psa_destroy_key( ctx->keys[1].key );
+
```
Comment (superna9999) [inline library/ssl_ticket.c:512]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:512]:
```suggestion
```

File: library/ssl_ticket.c [hunk 18]
```diff
@@ -418,9 +505,18 @@ int mbedtls_ssl_ticket_parse( void *p_ticket,
...
+#else
+
```
Comment (superna9999) [inline library/ssl_ticket.c:514]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:514]:
```suggestion
```

File: library/ssl_ticket.c [hunk 19]
```diff
@@ -418,9 +505,18 @@ int mbedtls_ssl_ticket_parse( void *p_ticket,
...
     mbedtls_cipher_free( &ctx->keys[0].ctx );
     mbedtls_cipher_free( &ctx->keys[1].ctx );
```
Comment (superna9999) [inline library/ssl_ticket.c:517]:
```suggestion
```

Comment (superna9999) [inline library/ssl_ticket.c:517]:
```suggestion
```

**Review by superna9999 — COMMENTED** [review]
Only spurious blank lines.

For the error handling, other SSL codes uses the following scheme:
```
        psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;

        if( ( status = psa_XXX( XXX ) ) != PSA_SUCCESS )
        {
            ret = psa_ssl_status_to_mbedtls( status );
            return( ret );
        }
```
I find it more readable, and would give coherency over the multiple SSL modules, but I don't have a strong opinion.

File: library/ssl_ticket.c [hunk 20]
```diff
@@ -177,29 +173,29 @@ int mbedtls_ssl_ticket_rotate( mbedtls_ssl_ticket_context *ctx,
...
                              PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT );
     psa_set_key_algorithm( &attributes, key->alg );
     psa_set_key_type( &attributes, key->key_type );
     psa_set_key_bits( &attributes, key->key_bits );
 
-    if( ( ret = psa_ssl_status_to_mbedtls(
-                    psa_import_key( &attributes, k,
-                                    PSA_BITS_TO_BYTES( key->key_bits ),
-                                    &key->key ) ) ) != 0 )
+    if( ( ret = psa_import_key( &attributes, k,
```
Comment (mpg) [inline library/ssl_ticket.c:188]: I think you mean `status` here.

**Review by mpg — CHANGES_REQUESTED** [review]
LGTM except for what I think is a typo, but unfortunately a significant one (variable name).

**Review by superna9999 — APPROVED** [review]

**Review by mpg — APPROVED** [review]
Thanks for fixing the typo I noticed, and also the other typo that I had missed! Looks all good to me now.

**Comment by mpg** [comment]
Regarding CI, the internal CI fully passed and the only two failures on the open CI are instances of #5583 and #5279 respectively, so they can be ignored as far as the present PR is concerned.

**Pull request merged**